### PR TITLE
Add module forces to trajectory states

### PR DESCRIPTION
**Pull request by BBScholar**
Addressing: https://github.com/SleipnirGroup/Choreo/issues/479  

General Thoughts/Questions:
- Do I need to think about the number of modules not being equal to 4? 

Todo: 
- [x] Ensure Serialization/Deserialization works for ChoreoLib (C++ and Java)

### CODE REVIEW

File: choreolib/src/main/native/include/choreo/lib/ChoreoTrajectoryState.h [hunk 1]
```diff
@@ -91,6 +99,12 @@ class ChoreoTrajectoryState {
   /// The angular component of the velocity at that point in the trajectory
   units::radians_per_second_t angularVelocity = 0_rad_per_s;
 
+  /// The forces on the modules in the X direction
+  std::array<units::newton_t, 4> moduleForcesX{{0_N, 0_N, 0_N, 0_N}};
+
+  /// The forces on the modules in the Y direction
+  std::array<units::newton_t, 4> moduleForcesY{{0_N, 0_N, 0_N, 0_N}};
```
Comment (spacey-sooty) [inline choreolib/src/main/native/include/choreo/lib/ChoreoTrajectoryState.h:106]: Do we wanna use `wpi::SmallSet` or something similar here?

File: choreolib/src/main/native/cpp/choreo/lib/ChoreoTrajectoryState.cpp [hunk 2]
```diff
@@ -49,14 +54,24 @@ frc::ChassisSpeeds ChoreoTrajectoryState::GetChassisSpeeds() const {
 ChoreoTrajectoryState ChoreoTrajectoryState::Interpolate(
     const ChoreoTrajectoryState& endValue, double i) const {
   frc::Pose2d interpPose = frc::Interpolate(GetPose(), endValue.GetPose(), i);
+
+  // TODO: Can this be done cleaner?
+  std::array<units::newton_t, 4> lerpFX, lerpFY;
+  for (int idx = 0; idx < 4; ++idx) {
+    lerpFX[idx] = wpi::Lerp(moduleForcesX[idx], endValue.moduleForcesX[idx], i);
+    lerpFY[idx] = wpi::Lerp(moduleForcesY[idx], endValue.moduleForcesY[idx], i);
+  }
```
Comment (spacey-sooty) [inline choreolib/src/main/native/cpp/choreo/lib/ChoreoTrajectoryState.cpp:63]: If we store these using an interpolatable type it could be cleaner.

File: choreolib/src/main/java/com/choreo/lib/ChoreoTrajectoryState.java [hunk 3]
```diff
@@ -91,14 +104,24 @@ public ChoreoTrajectoryState interpolate(ChoreoTrajectoryState endValue, double
     double scale = (t - this.timestamp) / (endValue.timestamp - this.timestamp);
     var interp_pose = getPose().interpolate(endValue.getPose(), scale);
 
+    // NOTE: Could maybe do this with streams? This seems more efficient for now.
+    double[] interp_fx = new double[4];
+    double[] interp_fy = new double[4];
+    for (int i = 0; i < 4; ++i) {
+      interp_fx[i] = MathUtil.interpolate(this.moduleForcesX[i], endValue.moduleForcesX[i], scale);
+      interp_fy[i] = MathUtil.interpolate(this.moduleForcesY[i], endValue.moduleForcesY[i], scale);
```
Comment (spacey-sooty) [inline choreolib/src/main/java/com/choreo/lib/ChoreoTrajectoryState.java:112]: If streams or some form of map could avoid allocating two new Arrays here that would be nice

**Review by spacey-sooty — COMMENTED** [review]
> Do I need to think about the number of modules not being equal to 4?
Does Trajoptlib/Choreo provide options to modify this? If so I would say yes

Does the C++ version compile? I've noticed a few places where units are passed into std functions which shouldn't work

**Comment by jlmcmchl** [comment]
Can this PR include documentation of the order of modules in the array? It's not obvious from this PR which is which.

File: choreolib/src/main/java/com/choreo/lib/ChoreoTrajectoryState.java (hunk 3 above)
Comment (calcmogul) [reply choreolib/src/main/java/com/choreo/lib/ChoreoTrajectoryState.java:112]: Streams don't necessarily avoid allocation. An array of primitives should end up being stack-allocated.

File: choreolib/src/main/native/include/choreo/lib/ChoreoTrajectoryState.h (hunk 1 above)
Comment (calcmogul) [reply choreolib/src/main/native/include/choreo/lib/ChoreoTrajectoryState.h:106]: Why?

File: choreolib/src/main/native/cpp/choreo/lib/ChoreoTrajectoryState.cpp (hunk 2 above)
Comment (calcmogul) [reply choreolib/src/main/native/cpp/choreo/lib/ChoreoTrajectoryState.cpp:63]: They're already inherently interpolatable. Locality of behavior is more important than abstraction.

File: choreolib/src/main/java/com/choreo/lib/ChoreoTrajectoryState.java (hunk 3 above)
Comment (BBScholar) [reply choreolib/src/main/java/com/choreo/lib/ChoreoTrajectoryState.java:112]: This was my thinking. The overhead of streams isn't worth it here IMO, especially given how simple the logic is.

**Comment by BBScholar** [comment]
> > Do I need to think about the number of modules not being equal to 4?
> 
> Does Trajoptlib/Choreo provide options to modify this? If so I would say yes
> 
> Does the C++ version compile? I've noticed a few places where units are passed into std functions which shouldn't work

I don't think Choreo itself supports more or less than 4 modules currently. Not sure about TrajoptLib, need to check.

This compiles on my machine (minus javadoc for some reason). Not sure why the units classes wouldn't work with standard lib. Are you talking about `std::transform` and similar?

**Comment by BBScholar** [comment]
For the document upgrade logic. Is it valid to add the module force arrays (zero initialized) and mark the paths as stale?

File: choreolib/src/main/native/cpp/choreo/lib/ChoreoTrajectoryState.cpp [hunk 4]
```diff
@@ -49,14 +53,24 @@ frc::ChassisSpeeds ChoreoTrajectoryState::GetChassisSpeeds() const {
...
       interpPose.Rotation().Radians(),
       wpi::Lerp(velocityX, endValue.velocityX, i),
       wpi::Lerp(velocityY, endValue.velocityY, i),
-      wpi::Lerp(angularVelocity, endValue.angularVelocity, i)};
+      wpi::Lerp(angularVelocity, endValue.angularVelocity, i),
+      lerpFX,
+      lerpFY};
 }
 
 std::array<double, 7> ChoreoTrajectoryState::AsArray() const {
```
Comment (BBScholar) [inline choreolib/src/main/native/cpp/choreo/lib/ChoreoTrajectoryState.cpp:76]: Should this include module data now? Leaning towards no.

Comment (calcmogul) [reply choreolib/src/main/native/cpp/choreo/lib/ChoreoTrajectoryState.cpp:76]: Idk what this function is even for, cuz we already support serialization a different way.

Comment (BBScholar) [reply choreolib/src/main/native/cpp/choreo/lib/ChoreoTrajectoryState.cpp:76]: I'll just leave it be then.

**Review by spacey-sooty — COMMENTED** [review]
Does as array need to be updated?

**Comment by calcmogul** [comment]
See https://github.com/SleipnirGroup/Choreo/pull/480/files/f2e212b53e282ce44c9706eff79e14cbbf854439#r1612273741.

**Comment by BBScholar** [comment]
Feels like that should be its own PR. But I can remove it here if that's best for you.

**Comment by calcmogul** [comment]
https://github.com/SleipnirGroup/Choreo/pull/482

**Comment by BBScholar** [comment]
Still need to check if (de)serialization works. Not sure if I should do that via unit tests or local test projects. Unit testing seems easier.

**Comment by BBScholar** [comment]
Wrote unit tests for json (de)serialization. Looks good. 

Tested the document upgrade logic on an old `.chor` file I had locally, looks good.

Should be ready for full review now.

**Review by spacey-sooty — APPROVED** [review]

File: src-tauri/tauri.conf.json [hunk 5]
```diff
@@ -60,7 +60,7 @@
       {
         "fullscreen": false,
         "resizable": true,
-        "title": "Choreo v2024.2.2",
+        "title": "Choreo v2024.3.0",
```
Comment (calcmogul) [inline src-tauri/tauri.conf.json:63]: This is typically done in a separate PR before release instead.

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Review by calcmogul — APPROVED** [review]

**Comment by calcmogul** [comment]
I'd expect them to be in the same coordinate system as the velocity.
